serveradmin/serverdb/migrations/0003_server_indexing.py

```diff
@@ -15,10 +21,10 @@ class Migration(migrations.Migration):
             '       intern_ip inet_ops WITH &&,'
             '       servertype_id WITH ='
             '   )'
+            # At InnoGames we have a servertype loadbalancer which is the only
+            # servertype allowed to have multiple objects sharing the same
+            # intern_ip.  We use this to forward different ports to different
+            # hosts as well as for multi datacenter virtual IPs.
             "   WHERE (servertype_id != 'loadbalancer')"
```

Comment:
This shouldn't be hardcoded, but use `ip_addr_type` of the servertype.  It is normally not allowed on Postgres for indexes to refer to other tables, but it'd work when the logic is wrapped into a function marked as `IMMUTABLE`.  The downside is that the changes on this column would corrupt the index.  I think it is unlikely for anybody to ever change `ip_addr_type` on an existing servertype.

Also the constraints can be made much better using such a function.  And I would also mark `ip_addr_type` field as not writable on Django admin at least to document this assumption.

Here is the SQL (not tested):

```sql
CREATE FUNCTION get_ip_addr_type(text)
    LANGUAGE sql STRICT IMMUTABLE
    AS 'SELECT ip_addr_type FROM servertype WHERE servertype_id = $1';

ALTER TABLE server
    ADD CONSTRAINT server_inter_ip_check
    CHECK ((get_ip_addr_type(servertype_id) = 'null') = (intern_ip IS NULL)),
    ADD CONSTRAINT server_host_inter_ip_exclude
    EXCLUDE (
        intern_ip WITH =
    )
    WHERE (get_ip_addr_type(servertype_id) = 'host'),
    ADD CONSTRAINT server_host_loadbalancer_inter_ip_exclude
    EXCLUDE (
        intern_ip WITH =,
        servertype_id WITH !=
    )
    WHERE (get_ip_addr_type(servertype_id) IN ('host', 'loadbalancer')),
    ADD CONSTRAINT server_network_inter_ip_exclude
    EXCLUDE USING gist (
        intern_ip inet_ops WITH &&,
        servertype_id WITH =
    )
    WHERE (get_ip_addr_type(servertype_id) = 'network');
```

Reply:
I agree, though this doesn't quite work yet.

I believe the function must define a return type like so:
```sql
CREATE FUNCTION get_ip_addr_type(text) RETURNS text
    LANGUAGE sql STRICT IMMUTABLE
    AS 'SELECT ip_addr_type FROM servertype WHERE servertype_id = $1';
```

The first two constraints seem fine, but the last two do not. I believe it's related to the inet datatype in combination with gist?

```
*serveradmin=# ALTER TABLE server
serveradmin-#     ADD CONSTRAINT server_host_loadbalancer_inter_ip_exclude
serveradmin-#     EXCLUDE (
serveradmin(#         intern_ip WITH =,
serveradmin(#         servertype_id WITH !=
serveradmin(#     )
serveradmin-#     WHERE (get_ip_addr_type(servertype_id) IN ('host', 'loadbalancer'));
ERROR:  operator <>(text,text) is not a member of operator family "text_ops"
DETAIL:  The exclusion operator must be related to the index operator class for the constraint.
```

```
*serveradmin=# ALTER TABLE server
serveradmin-#     ADD CONSTRAINT server_network_inter_ip_exclude
serveradmin-#     EXCLUDE USING gist (
serveradmin(#         intern_ip inet_ops WITH &&,
serveradmin(#         servertype_id WITH =
serveradmin(#     )
serveradmin-#     WHERE (get_ip_addr_type(servertype_id) = 'network');
ERROR:  data type character varying has no default operator class for access method "gist"
HINT:  You must specify an operator class for the index or define a default operator class for the data type.
```

Since this MR is only about cleaning up what's already there I think I will merge this so I can fake-migrate the whole thing once.

I'm interested in this change though and would like to keep track of it and create a new migration for it.

Reply:
`btree` is not supporting `<>` operator.  `btree_gist` extension should support.

Reply:
Ah, I see, thanks. So does this look good?

```sql
CREATE FUNCTION get_ip_addr_type(text) RETURNS text
    LANGUAGE sql STRICT IMMUTABLE
    AS 'SELECT ip_addr_type FROM servertype WHERE servertype_id = $1';

ALTER TABLE server
    ADD CONSTRAINT server_inter_ip_check
    CHECK ((get_ip_addr_type(servertype_id) = 'null') = (intern_ip IS NULL)),
    ADD CONSTRAINT server_host_inter_ip_exclude
    EXCLUDE (
        intern_ip WITH =
    )
    WHERE (get_ip_addr_type(servertype_id) = 'host'),
    ADD CONSTRAINT server_host_loadbalancer_inter_ip_exclude
    EXCLUDE USING gist (
        intern_ip WITH =,
        servertype_id WITH !=
    )
    WHERE (get_ip_addr_type(servertype_id) IN ('host', 'loadbalancer')),
    ADD CONSTRAINT server_network_inter_ip_exclude
    EXCLUDE USING gist (
        intern_ip inet_ops WITH &&,
        servertype_id WITH =
    )
    WHERE (get_ip_addr_type(servertype_id) = 'network');

ALTER TABLE server DROP CONSTRAINT server_inter_ip_exclude;
```

Reply:
Applied, seems to work fine.